### PR TITLE
Extend export and cache image

### PR DESCRIPTION
**Pull request by hawkeyexp**
Cache an image as thumbnail which is shown in exported listing. Looks better :-)
I also tried to grab the image from kodi db (if kodi has added export to db)  via rpc but it was to slow for my feeling.

### CODE REVIEW

**Comment by hawkeyexp** [comment]
Wait with merge - try to fix the problem of milhouse to complete in one shot.

**Comment by hawkeyexp** [comment]
it seems xbmcvfs.exists is part of the problem. On nfs it works - on local storage and smb it fails.

**Comment by MilhouseVH** [comment]
Are the failing paths included in your `sources.xml`? Kodi will only access content that is in `sources.xml`.

**Comment by hawkeyexp** [comment]
@MilhouseVH 
the interesting part is that it can write with xbmcvfs without problems only xbmcvfs.exists fails - is this normal or possible a bug? It makes no sense for me that writing works without part of sources.xml but xbmcvfs.exists fails. Possible it would be great if all parts of xbmcvfs works without limit to be part of sources.xml.

**Comment by MilhouseVH** [comment]
That I don't know - I'm only vaguely aware that Kodi won't (or should not) read content that is not referenced within `sources.xml`. This is a very basic security mechanism. However I wouldn't be at all surprised if there's a code path, such as writing, which manages (most likely unintentionally) to avoid the security check.

It might be worth opening a ticket on trac.kodi.tv with a simple example demonstrating the write issue, then the various Kodi developers familiar with the xbmcvfs (and also the "security mechanism") can wade in.

**Comment by asciidisco** [comment]
Using exists for file system operations is considered a bad practice anyway (I know I'm guilty of this myself, because sometimes we're all lazy).
It's basically a built in invitation for race conditions. If exists is really the problem, just use read in a try/except block.

**Comment by hawkeyexp** [comment]
I played arround and there must be really a bug in xbmcvfs. It makes no differ if export folder is in sources.xml or not. Currently it only works well on nfs shares. I think we should merge the latest commits to prevent from bugs (empty folder is shown if something goes wrong) but i will open an issue for kodi. For my understanding xbmcvfs should solve problems with python os module on different destination types but it's currently buggy.

**Comment by hawkeyexp** [comment]
Ticked is opened on trac.kodi.tv (#17606)

**Comment by MilhouseVH** [comment]
In https://trac.kodi.tv/ticket/17606 you should probably include links to the relevant PRs in this repo.

**Comment by hawkeyexp** [comment]
Finally i could identify the exact problem! xbmcvfs.exists needs a final path delimator at the end of string to work correct with folders (on files it doesn't need it). I added a check for folderpaths.
@MilhouseVH can you try it out - for me it is working without errors now in nfs smb and local storage.

With a new pull request i will add a check if destination is listed in soruces.xml and optional auto add it like sandmann79 does it in his amazon addon.

**Comment by MilhouseVH** [comment]
Yes that seems to work - `xbmcvfs.exists(kodi_helper.check_folder_path(self.movie_path))` and `xbmcvfs.exists(kodi_helper.check_folder_path(self.tvshow_path))` both now return `True` for me.

Just one small nit with this PR - you are introducing some whitespace changes which you might want to avoid:
s1

File: resources/lib/KodiHelper.py [hunk 1]
```diff
@@ -623,7 +623,7 @@ def build_video_listing_exported (self, content, build_url):
             })
             li.setArt(art)
             xbmcplugin.addDirectoryItem(handle=self.plugin_handle, url=url, listitem=li, isFolder=isFolder)
```
Comment (MilhouseVH) [inline resources/lib/KodiHelper.py]: ^^^ One more at the end of the line. :)

**Comment by hawkeyexp** [comment]
Cleanup and rework patch

**Comment by MilhouseVH** [comment]
Your editor might have a configuration setting to control whitespace, such as "Strip trailing spaces from lines when saving" (or similar), which you might want to consider enabling.

**Comment by hawkeyexp** [comment]
Normally i edit in a linux mashine but this change was made on a win mashine and forgot to check it :-)

**Comment by asciidisco** [comment]
@kat953162 Can you provide a debug log?

**Comment by kat953162** [comment]
11:06:37.480 T:9884   ERROR: CAddonMgr: failed to unload plugin.video.netflix
11:06:37.480 T:9884   ERROR: CAddonInstallJob[plugin.video.netflix]: failed to unload addon.

I will try with a different install and get you a full log if it doesn't work. Meanwhile, this dependency shows as Unavailable: https://puu.sh/xIxXe/91ceebc5e6.png
I tried manually installing the dependency because I couldn't get the addon to reinstall.

**Comment by kat953162** [comment]
I noticed this in my log...

11:21:54.661 T:11260   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.ImportError'>
                                            Error Contents: No module named Cryptodome.PublicKey
                                            Traceback (most recent call last):
                                              File "C:\Users\Azriel\AppData\Roaming\Kodi\addons\plugin.video.netflix-master\addon.py", line 7, in <module>
                                                from resources.lib.KodiHelper import KodiHelper
                                              File "C:\Users\Azriel\AppData\Roaming\Kodi\addons\plugin.video.netflix-master\resources\lib\KodiHelper.py", line 11, in <module>
                                                from MSL import MSL
                                              File "C:\Users\Azriel\AppData\Roaming\Kodi\addons\plugin.video.netflix-master\resources\lib\MSL.py", line 20, in <module>
                                                from Cryptodome.PublicKey import RSA
                                            ImportError: No module named Cryptodome.PublicKey
                                            -->End of Python script error report<--

**Comment by kat953162** [comment]
Ignore what I said earlier, clearing kodi info got it to install this time...
I'm having this issue, trying different nightlies:
https://github.com/asciidisco/plugin.video.netflix/issues/85

**Comment by hawkeyexp** [comment]
@MilhouseVH 

Sorry to go this way but i have no other way to cotanct you :-(
Is there any way to manually set the tvshowid for current playing media ? I try to fix the not working nextup service the best way and this is to deliver this id - i can get it from kodi db this is no prob but i need to set it inside the payer but dbid (only possible value) is not setting tvshowid.

**Comment by MilhouseVH** [comment]
@hawkeyexp Sorry I've no idea really, I only have a basic understanding of the JSON API and this probably isn't of any use to you - you probably need to speak with someone who knows about the Kodi addon system... hopefully the forums will return sooner than later.

**Pull request merged**